src/dvc_data/hashfile/checkout.py

```diff
@@ -114,6 +118,52 @@ def _checkout_file(
     return modified
 
 
+def _needs_relink(
+    path: str, cache: "HashFileDB", meta: "Meta", oid: Optional[str]
+) -> bool:
+    destination = meta.destination
+    is_symlink = meta.is_link
+    is_hardlink = meta.nlink > 1
+    is_copy = not is_symlink and not is_hardlink
+
+    obj_path: Optional[str] = None
+
+    for link_type in cache.cache_types:
+        if link_type in ("copy", "reflink") and is_copy:
+            return False
+        if not oid:
+            continue
+
+        if obj_path is None:
+            obj_path = cache.oid_to_path(oid)
+
+        if link_type == "symlink" and is_symlink and destination:
+            return destination != obj_path
+        if link_type == "hardlink" and is_hardlink and samefile(path, obj_path):
```

Comment:
This os call is still a bottleneck. Here's the profile for the 1.5 million files in the [imagenet 2017 annotations](https://image-net.org/data/bboxes_annotations.tar.gz):

[add-relink-diff.prof.zip](https://github.com/user-attachments/files/16552039/add-relink-diff.prof.zip)

For this to be really worth it, I think we would at least have to parallelize so these are less of a blocker or else rely on metadata to determine if they are the same.

Reply:
Opened https://github.com/iterative/dvc-data/pull/549 that adds a thread pool. Feel free to reimplement a different way if you want, but this has a much more reasonable profile:

[add-relink-diff-thread.prof.zip](https://github.com/user-attachments/files/16553049/add-relink-diff-thread.prof.zip)

Reply:
> Here's the profile for the 1.5 million files 

I downloaded that, extracted that file. There were `tar.gz` files inside again, which I re-extracted. And I deleted the existing `.tar.gz` file.

I see following stats that says 1.07 million files. Did I do anything wrong?
```
Number of files: 1,077,367 (reg: 1,073,739, dir: 3,628)
```

---


I know macOS filesystem operations are slow, but I did not know they were this terrible. 
On my machine (Linux/Intel), `_check_relink` finishes in 5s - 6s. First time I did `dvc add`, it took ~9mins, and repeated `dvc add` takes <1min 30s.

I can drop one more stat calls and use metadata. But I need to stat the path to object in the cache to compare inodes (we don't have cache metadata). That drops to 3s for me. :)

Reply:
Btw, I changed this to be strictly metadata comparison now. The `inode` data comes from when we check for object integrity with `HashFileDB.check()` (it's a bit hacky but works).

Reply:
> On my machine (Linux/Intel), `_check_relink` finishes in 5s - 6s. First time I did `dvc add`, it took ~9mins, and repeated `dvc add` takes <1min 30s.

🤯 I just spun up a linux vm and still got 40+ minutes for the same files with this PR. I guess it shows how much the filesystem impacts performance, but I think you have done what you can with this PR.

Reply:
Could you please share how much time `_determine_files_to_relink` takes on either machines? I am just curious what it costs.

You can add `@funcy.print_durations` at the top of the function, and then `dvc add ...`.

---

> I guess it shows how much the filesystem impacts performance


This is what I see in my machine for `posix.stat`.

ncalls | tottime | percall | cumtime | percall | filename:lineno(function)
-- | -- | -- | -- | -- | --
354974 | 0.9792 | 2.758e-06 | 0.9792 | 2.758e-06 | ~:0(\<built-in method posix.stat\>)


Compared that to what I see in yours:


ncalls | tottime | percall | cumtime | percall | filename:lineno(function)
-- | -- | -- | -- | -- | --
11907017 | 2235 | 0.0001877 | 2235 | 0.0001877 | ~:0(\<built-in method posix.stat\>)


which is 68x slower.

Reply:
> Could you please share how much time `_determine_files_to_relink` takes on either machines? I am just curious what it costs.

The last run I did on my mac with this PR, this only took ~4 seconds. Most of the time is spent on stat calls as you noted.